### DISCUSSION TITLE
coupling of TH and poroelastic mechanics (HM)

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#21062] [Q&A Modules: Solid mechanics] [answered]
Hello everyone, 

I meet a poroelastic problem of coupled natural convection (TH) and poroelastic mechanics (HM). 

This benchmark is simplified based on this work using Comsol (https://www.essoar.org/pdfjs/10.1002/essoar.10508352.1).

Problem information:
Aim: To study the effect of the mechanic (M) on the buoyancy-driven convection (TH)
Geometry: x and y is (0, 5500) m; z is (-5500, 0) m; Fig. 1.
Blocks: basement (impermeable) and fault (permeable)
BCs: (1) Top surface: fixed 293.15 K temperature and 10^5 Pa porepressure; (2) Bottom surface: NeummannBC of temperature is 0.1 W/m^2, disp_x,y,z =0;
ICs: (1) temperature: '293.15-z*0.03' ; (2) porepressure: '100000-1000*9.81*z'
EOS: same to the above paper. Water density is the function of temperature and porepressure. Water viscosity is the function of temperature.
permeability and porosity evolution: no.
 automatic_scaling = true
Couplings: TH (i.e. buoyancy-driven convection) and HM (i.e. poroelastic) couplings. No thermal expansion and volume expansion.

Forces: only body force (gravity).
Other detailed properties are shown in input files.

**My work:** 

1) buoyancy-driven convection (TH).

The result is Ok.

2) HM poroelastic mechanics

I test some examples from (https://github.com/ajacquey/golem/tree/master/test/tests/HM). The result is also fine.

3) coupling of buoyancy-driven convection (TH) and poroelastic mechanics (HM)

(1) Steady simulation; 
kernels: 
- pure heat conduction (T kernel)
- pure porepressure diffusion (H)
- body force kernel (my app)
- PoroMechanicsCoupling
- Tensor Mechanics Master Action System (M kernels).
EOS: constant water density 1000 (kg/m3)and viscosity (1e-3 Pa.s).

(2) Transient simulation: 
- the kernels in steady simulation
- transient temperature and pore pressure kernels
- HM coupling kernel (my app, _biot[_qp] * _vol_strain_rate[_qp] * _test[_i][_qp])

a. It uses the output results (temperature, pore pressure, displacements (using or ignoring)) of steady simulation as the initial values of Transient simulation. 

b. Moreover, I learn the examples about setting 'ComputeEigenstrainFromInitialStress' (regarding body force as initial stress) from https://github.com/idaholab/falcon/blob/devel/test/tests/THM_injection/thm_steady.i and https://groups.google.com/g/moose-users/c/ceMFbx9CgiE. To be honest, I am not sure whether should I regarding body force as initial stress.

c. The application of the stresses is applied progressively by setting the body gravity increases from 0.01 to 9.81 in 10 years.

However, all my previous Transient simulations output very terrible convergence information (as follows) even at first time step t=1s.

**My request: Could you please give me some tips on my problem? Thanks a lot!**

_**good steady simulations information**_

```
 0 Nonlinear |R| = [32m7.435105e+00[39m
    |residual|_2 of individual variables:
                disp_x:      0.795433
                disp_y:      1.20725
                disp_z:      0.670134
                pressure:    4.87658e-08
                temperature: 7.26234
    |residual|_2 of individual variables:
                disp_x:      0.795433
                disp_y:      1.20725
                disp_z:      0.670134
                pressure:    4.87658e-08
                temperature: 7.26234
      0 Linear |R| = [32m7.435105e+00[39m
      1 Linear |R| = [32m3.257785e+00[39m
      2 Linear |R| = [32m1.109984e+00[39m
...
     26 Linear |R| = [32m1.102510e-07[39m
     27 Linear |R| = [32m4.116438e-08[39m
  Linear solve converged due to CONVERGED_RTOL iterations 27
 1 Nonlinear |R| = [32m9.905635e-03[39m
    |residual|_2 of individual variables:
                disp_x:      1.856e-08
                disp_y:      2.33502e-08
                disp_z:      0.0040312
                pressure:    4.44167e-08
                temperature: 0.00904826
    |residual|_2 of individual variables:
                disp_x:      1.856e-08
                disp_y:      2.33502e-08
                disp_z:      0.0040312
                pressure:    4.44167e-08
                temperature: 0.00904826
      0 Linear |R| = [32m9.905635e-03[39m
...
     27 Linear |R| = [32m4.733500e-11[39m
  Linear solve converged due to CONVERGED_RTOL iterations 27

 4 Nonlinear |R| = [32m4.264257e-08[39m
    |residual|_2 of individual variables:
                disp_x:      8.95223e-15
                disp_y:      7.12969e-15
                disp_z:      1.21779e-10
                pressure:    4.26423e-08
                temperature: 1.08163e-10
    |residual|_2 of individual variables:
                disp_x:      8.95223e-15
                disp_y:      7.12969e-15
                disp_z:      1.21779e-10
                pressure:    4.26423e-08
                temperature: 1.08163e-10
Nonlinear solve converged due to CONVERGED_FNORM_RELATIVE iterations 4
[32m Solve Converged![39m

Outlier Variable Residual Norms:
  pressure: [31m4.264226e-08[39m
```

_**bad transient simulations information**_

```
Time Step 0, time = 0
Outputting exodus ....                                                                     [[33m  5.95 s[39m]

Time Step 1, time = 1, dt = 1
Computing initial residual ......                                                          [[33m  7.35 s[39m]

Performing automatic scaling calculation

 0 Nonlinear |R| = [32m8.372612e+07[39m
    |residual|_2 of individual variables:
                disp_x:      1.81623e-07
                disp_y:      6.55306e-08
                disp_z:      7.96715e-08
                pressure:    8.37261e+07
                temperature: 1.43173e-07
  0 SNES Function norm 8.372611823977e+07 
Building node to element map                                                               [[33m  1.42 s[39m]
      0 Linear |R| = [32m8.372612e+07[39m
      1 Linear |R| = [32m3.617225e+03[39m
  Linear solve converged due to CONVERGED_RTOL iterations 1
 1 Nonlinear |R| = [32m3.040338e+07[39m
    |residual|_2 of individual variables:
                disp_x:      9.35686e-05
                disp_y:      7.84779e-05
                disp_z:      0.000113443
                pressure:    3.04034e+07
                temperature: 1.71846e-08
  1 SNES Function norm 3.040338271683e+07 
      0 Linear |R| = [32m3.040338e+07[39m
      1 Linear |R| = [32m1.046802e+03[39m
  Linear solve converged due to CONVERGED_RTOL iterations 1
...
25 Nonlinear |R| = [32m6.088715e+04[39m
    |residual|_2 of individual variables:
                disp_x:      2.60351e-07
                disp_y:      2.47446e-07
                disp_z:      1.93876e-07
                pressure:    60887.2
                temperature: 2.08727e-10
 25 SNES Function norm 6.088715233063e+04 
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 25
[31m Solve Did NOT Converge![39m
Aborting as solve did not converge

Solve failed, cutting timestep.
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

A few questions:
- is the gravity ramp up consistent between the steady and transient solutions? Or is the pressure field initialization for the real gravity in the transient?
- does it converge without the coupling term, with just the addition of the time derivatives
- did you implement the Jacobian term for your custom coupling kernel? Or are you using automatic differentiation?

Guillaume
